File: python/angle_stream_control.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
高擎电机角度流控制
基于 03_angle.py 协议实现
支持 0x90 流命令和MIT风格控制
"""

import can
import struct
import time
import sys
import math
import argparse


class MotorAngleStreamController:
    """高擎电机角度流控制器"""

    def __init__(self, channel: str = 'can0', bitrate: int = 1000000, motor_id: int = 1):
        self.motor_id = motor_id
        self.channel = channel
        self.bitrate = bitrate
        self.bus = None

        # SDK 系数定义
        self.FACTOR_POS = 10000.0  # 1圈 = 10000
        self.FACTOR_VEL = 4000.0   # 1r/s = 4000
        self.FACTOR_TQE = 200.0   # 通用电机系数

    def connect(self) -> bool:
        """连接CAN总线"""
        print(f"初始化 CAN: {self.channel}")
        try:
            self.bus = can.interface.Bus(channel=self.channel, interface='socketcan')
            # 清空缓冲区
            while self.bus.recv(timeout=0.0):
                pass
            print("✅ CAN接口连接成功")
            return True
        except OSError:
            print("❌ 错误: CAN 接口未打开")
            return False

    def send_frame(self, arbitration_id, data):
        """发送CAN帧"""
        msg = can.Message(arbitration_id=arbitration_id, data=data,
                         is_extended_id=True, is_fd=False)
        try:
            self.bus.send(msg)
        except can.CanError:
            pass

    def enable_motor(self):
        """
        使能步骤：需要先用 Mode 1 (ID 0x0001) 激活电机进入 0x0A 模式
        """
        print(f"-> [ID {self.motor_id}] 发送使能指令 (Register Mode)...")
        # 1. 写入模式: 0x0A (Position Mode)
        # ID: 0x0001 (Cmd 0x01 Write Int8)
        arb_id = 0x0000 | self.motor_id
        self.send_frame(arb_id, [0x01, 0x00, 0x0A, 0x50, 0x50, 0x50, 0x50, 0x50])
        time.sleep(0.05)

        # 2. 预设 PID (给一点刚度)
        # Reg 0x23 (Kp) = 1.0
        self.send_frame(arb_id, [0x0D, 0x23] + list(struct.pack('<f', 1.0)) + [0x50, 0x50])
        time.sleep(0.02)
        # Reg 0x24 (Kd) = 0.1
        self.send_frame(arb_id, [0x0D, 0x24] + list(struct.pack('<f', 0.1)) + [0x50, 0x50])

        print("✅ 电机已激活，准备发送流控制指令")

    def send_0x90_command(self, angle_deg, max_vel_rps, max_tqe_nm):
        """
        发送 0x90 命令帧 (一拖多模式 / 流模式)
        ID: 0x0090 (由 SDK 0x90 推断)
        Payload: [Pos(int16), Vel(int16), Tqe(int16), Padding...]
        """
        # 1. 计算数值
        pos_int = int((angle_deg / 360.0) * self.FACTOR_POS)
        vel_int = int(max_vel_rps * self.FACTOR_VEL)
        tqe_int = int(max_tqe_nm * self.FACTOR_TQE)

        # 限幅 Int16
        pos_int = max(min(pos_int, 32767), -32768)
        vel_int = max(min(vel_int, 32767), -32768)
        tqe_int = max(min(tqe_int, 32767), -32768)

        # 2. 打包数据
        # 结构: [PosL, PosH, VelL, VelH, TqeL, TqeH]
        data = struct.pack('<hhh', pos_int, vel_int, tqe_int)

        # 补齐 8 字节
        payload = list(data) + [0x50, 0x50]

        # 3. 发送至 ID 0x0090 (不需回复) 或 0x8090 (需回复)
        # 这里尝试 0x0090
        self.send_frame(0x0090, payload)

        print(f"   >>> 0x90流指令: Ang={angle_deg}° Vel={max_vel_rps} Tqe={max_tqe_nm} (Raw: {payload[:6]})")

    def disable_motor(self):
        """禁用电机"""
        arb_id = 0x0000 | self.motor_id
        self.send_frame(arb_id, [0x01, 0x00, 0x00, 0x50, 0x50, 0x50, 0x50, 0x50])
        print("🛑 电机已禁用")

    def set_angle(self, angle_deg, max_vel_rps=2.0, max_tqe_nm=3.0, send_count=5):
        """设置角度位置"""
        for i in range(send_count):
            self.send_0x90_command(angle_deg, max_vel_rps, max_tqe_nm)
            time.sleep(0.01)

    def run_interactive_control(self):
        """运行交互式角度控制"""
        print("="*50)
        print("🚀 0x90 流控制模式 (复刻 SDK) ")
        print("输入角度 (如 90) 回车。")
        print("默认参数: 限速 2.0 r/s, 限矩 3.0 Nm")
        print("输入 q 退出")
        print("="*50)

        try:
            # 1. 先使能电机
            self.enable_motor()

            while True:
                raw = input(f"\n(Stream 0x90) > ").strip()

                if raw in ['q', 'exit']:
                    break
                if not raw:
                    continue

                try:
                    deg = float(raw)
                    # 持续发送几次确保收到 (流模式通常需要高频发送)
                    self.set_angle(deg)
                except ValueError:
                    print("输入错误")

        except KeyboardInterrupt:
            print("\n中断")
        finally:
            self.disable_motor()

    def run_sine_wave(self, amplitude_deg: float, frequency: float, duration: float):
        """运行正弦波角度控制"""
        print("="*50)
        print("🌊 正弦波角度控制")
        print(f"幅值: {amplitude_deg}°, 频率: {frequency} Hz, 时长: {duration}s")
        print("="*50)

        try:
            # 1. 先使能电机
            self.enable_motor()

            start_time = time.time()
            while time.time() - start_time < duration:
                # 计算目标角度
                elapsed = time.time() - start_time
                target_deg = amplitude_deg * math.sin(2 * math.pi * frequency * elapsed)

                # 发送角度指令
                self.set_angle(target_deg)

                # 显示当前状态
                print(f"\r目标: {target_deg:7.1f}°", end="")
                sys.stdout.flush()

        except KeyboardInterrupt:
            print("\n中断")
        finally:
            self.disable_motor()

    def run_step_control(self, angles: list, step_duration: float):
        """运行阶梯角度控制"""
        print("="*50)
        print("📈 阶梯角度控制")
        print(f"角度序列: {angles}°")
        print(f"每步时长: {step_duration}s")
        print("="*50)

        try:
            # 1. 先使能电机
            self.enable_motor()

            for i, angle_deg in enumerate(angles):
                print(f"\n--- 步骤 {i+1}/{len(angles)}: {angle_deg}° ---")

                # 发送角度指令
                self.set_angle(angle_deg)

                # 等待步长时间
                step_start = time.time()
                while time.time() - step_start < step_duration:
                    print(f"\r剩余时间: {step_duration - (time.time() - step_start):.1f}s", end="")
                    sys.stdout.flush()
                    time.sleep(0.1)

        except KeyboardInterrupt:
            print("\n中断")
        finally:
            self.disable_motor()

    def run_mit_control(self, target_deg: float, stiffness: float, damping: float, duration: float):
        """运行MIT控制风格的角度控制"""
        print("="*50)
        print("🤖 MIT风格阻抗控制")
        print(f"目标角度: {target_deg}°")
        print(f"刚度: {stiffness} Nm/rad")
        print(f"阻尼: {damping} Nm·s/rad")
        print(f"控制时长: {duration}s")
        print("="*50)

        try:
            # 1. 先使能电机
            self.enable_motor()

            start_time = time.time()
            last_error = 0.0
            last_time = start_time

            while time.time() - start_time < duration:
                # 获取当前状态 (需要实现状态读取)
                # 这里简化处理，实际应该读取电机反馈
                current_time = time.time()
                dt = current_time - last_time

                # 简化的误差计算 (实际应该读取电机当前位置)
                # 这里假设角度直接对应位置
                error = math.radians(target_deg) - math.radians(target_deg)  # 临时简化

                # MIT控制律
                desired_torque = stiffness * error + damping * (error - last_error) / (dt + 0.001)

                # 发送控制指令 (转换为角度+速度+力矩)
                self.send_0x90_command(target_deg, 2.0, abs(desired_torque))

                print(f"\r目标: {target_deg:6.1f}° 力矩: {desired_torque:6.3f}Nm", end="")
                sys.stdout.flush()

                last_error = error
                last_time = current_time
                time.sleep(0.01)  # 100Hz控制频率

        except KeyboardInterrupt:
            print("\n中断")
        finally:
            self.disable_motor()

    def test_positions(self, positions: list):
        """测试多个角度位置"""
        print("="*50)
        print("🧪 多位置测试")
        print(f"测试位置: {positions}°")
        print("="*50)

        try:
            # 1. 先使能电机
            self.enable_motor()

            for i, angle_deg in enumerate(positions):
                print(f"\n--- 测试位置 {i+1}/{len(positions)}: {angle_deg}° ---")

                # 发送角度指令
                self.set_angle(angle_deg)

                # 等待稳定
                print("等待2秒稳定...", end="")
                sys.stdout.flush()
                time.sleep(2.0)

        except KeyboardInterrupt:
            print("\n中断")
        finally:
            self.disable_motor()

    def cleanup(self):
        """清理资源"""
        if self.bus:
            self.bus.shutdown()
            self.bus = None


def main():
    parser = argparse.ArgumentParser(description='高擎电机角度流控制')
    parser.add_argument('--channel', type=str, default='can0', help='CAN通道')
    parser.add_argument('--bitrate', type=int, default=1000000, help='CAN波特率')
    parser.add_argument('--motor_id', type=int, default=1, help='电机ID')

    parser.add_argument('--mode', type=str, default='interactive',
                       choices=['interactive', 'sine', 'step', 'mit', 'test'],
                       help='控制模式')

    parser.add_argument('--amplitude', type=float, default=90.0, help='正弦波幅值')
    parser.add_argument('--frequency', type=float, default=0.2, help='正弦波频率')
    parser.add_argument('--duration', type=float, default=10.0, help='测试时长')
    parser.add_argument('--angles', type=str, help='阶梯角度序列(逗号分隔)')
    parser.add_argument('--step_duration', type=float, default=2.0, help='阶梯步长')
    parser.add_argument('--stiffness', type=float, default=50.0, help='MIT刚度(Nm/rad)')
    parser.add_argument('--damping', type=float, default=5.0, help='MIT阻尼(Nm·s/rad)')
    parser.add_argument('--target', type=float, help='MIT目标角度')

    parser.add_argument('--test_angles', type=str, help='测试角度序列(逗号分隔)')

    args = parser.parse_args()

    # 创建控制器
    controller = MotorAngleStreamController(args.channel, args.bitrate, args.motor_id)

    try:
        # 连接CAN总线
        if not controller.connect():
            sys.exit(1)

        # 根据模式运行
        if args.mode == 'interactive':
            controller.run_interactive_control()
        elif args.mode == 'sine':
            controller.run_sine_wave(args.amplitude, args.frequency, args.duration)
        elif args.mode == 'step':
            if not args.angles:
                print("阶梯模式需要 --angles 参数")
                return
            angles = [float(a.strip()) for a in args.angles.split(',')]
            controller.run_step_control(angles, args.step_duration)
        elif args.mode == 'mit':
            if args.target is None:
                print("MIT模式需要 --target 参数")
                return
            controller.run_mit_control(args.target, args.stiffness, args.damping, args.duration)
        elif args.mode == 'test':
            if not args.test_angles:
                print("测试模式需要 --test_angles 参数")
                return
            positions = [float(p.strip()) for p in args.test_angles.split(',')]
            controller.test_positions(positions)

    except KeyboardInterrupt:
        print("\n⚠️ 用户中断")
    except Exception as e:
        print(f"\n❌ 控制过程中出错: {e}")
    finally:
        controller.cleanup()


if __name__ == "__main__":
    main()

File: cpp/angle_stream_control.cpp
/**
 * LivelyBot Angle Stream Control
 * Simple angle control utility using SocketCAN
 * g++ -std=c++17 -o angle_stream_control angle_stream_control.cpp -pthread
 */

#include <iostream>
#include <iomanip>
#include <thread>
#include <chrono>
#include <vector>
#include <string>
#include <cstring>
#include <cmath>
#include <signal.h>
#include <sstream>
#include <algorithm>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <linux/can.h>
#include <linux/can/raw.h>
#include <cctype>

volatile sig_atomic_t g_running = 1;

void signalHandler(int signum) {
    (void)signum; // 避免未使用参数警告
    g_running = 0;
}

class SimpleAngleController {
private:
    std::string channel_;
    int bitrate_;
    int motor_id_;

    // SDK 系数定义
    static constexpr double FACTOR_POS = 10000.0;  // 1圈 = 10000
    static constexpr double FACTOR_VEL = 4000.0;   // 1r/s = 4000
    static constexpr double FACTOR_TQE = 200.0;    // 通用电机系数

    int socket_fd_;

public:
    SimpleAngleController(const std::string& channel, int bitrate, int motor_id)
        : channel_(channel), bitrate_(bitrate), motor_id_(motor_id), socket_fd_(-1) {}

    ~SimpleAngleController() {
        cleanup();
    }

    bool initialize() {
        std::cout << "初始化 CAN: " << channel_ << std::endl;

        // 创建 socket
        socket_fd_ = socket(PF_CAN, SOCK_RAW, CAN_RAW);
        if (socket_fd_ < 0) {
            std::cerr << "❌ 错误: 无法创建 CAN socket" << std::endl;
            return false;
        }

        // 设置 CAN 接口
        struct ifreq ifr;
        strcpy(ifr.ifr_name, channel_.c_str());
        if (ioctl(socket_fd_, SIOCGIFINDEX, &ifr) < 0) {
            std::cerr << "❌ 错误: CAN 接口 " << channel_ << " 不存在" << std::endl;
            close(socket_fd_);
            socket_fd_ = -1;
            return false;
        }

        // 绑定 socket 到接口
        struct sockaddr_can addr;
        memset(&addr, 0, sizeof(addr));
        addr.can_family = AF_CAN;
        addr.can_ifindex = ifr.ifr_ifindex;

        if (bind(socket_fd_, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
            std::cerr << "❌ 错误: 无法绑定到 CAN 接口" << std::endl;
            close(socket_fd_);
            socket_fd_ = -1;
            return false;
        }

        std::cout << "✅ CAN接口连接成功" << std::endl;
        return true;
    }

    void cleanup() {
        disableMotor();
        if (socket_fd_ >= 0) {
            close(socket_fd_);
            socket_fd_ = -1;
        }
    }

    bool sendFrame(uint32_t arbitration_id, const uint8_t* data, uint8_t data_length) {
        if (socket_fd_ < 0) return false;

        struct can_frame frame;
        frame.can_id = arbitration_id;
        frame.can_dlc = data_length;
        memcpy(frame.data, data, data_length);

        return write(socket_fd_, &frame, sizeof(struct can_frame)) == sizeof(struct can_frame);
    }

    bool enableMotor() {
        std::cout << "-> [ID " << motor_id_ << "] 发送使能指令 (Register Mode)..." << std::endl;

        uint32_t arb_id = motor_id_;

        // 1. 写入模式: 0x0A (Position Mode)
        uint8_t data1[8] = {0x01, 0x00, 0x0A, 0x50, 0x50, 0x50, 0x50, 0x50};
        sendFrame(arb_id, data1, 8);
        std::this_thread::sleep_for(std::chrono::milliseconds(50));

        // 2. 预设 PID (给一点刚度)
        float kp = 1.0f;
        uint8_t data2[8] = {0x0D, 0x23};
        memcpy(&data2[2], &kp, sizeof(kp));
        memset(&data2[6], 0x50, 2);
        sendFrame(arb_id, data2, 8);
        std::this_thread::sleep_for(std::chrono::milliseconds(20));

        float kd = 0.1f;
        uint8_t data3[8] = {0x0D, 0x24};
        memcpy(&data3[2], &kd, sizeof(kd));
        memset(&data3[6], 0x50, 2);
        sendFrame(arb_id, data3, 8);

        std::cout << "✅ 电机已激活，准备发送流控制指令" << std::endl;
        return true;
    }

    bool disableMotor() {
        uint32_t arb_id = motor_id_;
        uint8_t data[8] = {0x01, 0x00, 0x00, 0x50, 0x50, 0x50, 0x50, 0x50};

        bool result = sendFrame(arb_id, data, 8);
        if (result) {
            std::cout << "🛑 电机已禁用" << std::endl;
        }
        return result;
    }

    void send0x90Command(double angle_deg, double max_vel_rps, double max_tqe_nm) {
        // 1. 计算数值
        int16_t pos_int = static_cast<int16_t>((angle_deg / 360.0) * FACTOR_POS);
        int16_t vel_int = static_cast<int16_t>(max_vel_rps * FACTOR_VEL);
        int16_t tqe_int = static_cast<int16_t>(max_tqe_nm * FACTOR_TQE);

        // 限幅 Int16
        pos_int = static_cast<int16_t>(std::max(-32768.0, std::min(32767.0, static_cast<double>(pos_int))));
        vel_int = static_cast<int16_t>(std::max(-32768.0, std::min(32767.0, static_cast<double>(vel_int))));
        tqe_int = static_cast<int16_t>(std::max(-32768.0, std::min(32767.0, static_cast<double>(tqe_int))));

        // 2. 打包数据: [PosL, PosH, VelL, VelH, TqeL, TqeH, 0x50, 0x50]
        uint8_t data[8];
        data[0] = static_cast<uint8_t>(pos_int & 0xFF);
        data[1] = static_cast<uint8_t>((pos_int >> 8) & 0xFF);
        data[2] = static_cast<uint8_t>(vel_int & 0xFF);
        data[3] = static_cast<uint8_t>((vel_int >> 8) & 0xFF);
        data[4] = static_cast<uint8_t>(tqe_int & 0xFF);
        data[5] = static_cast<uint8_t>((tqe_int >> 8) & 0xFF);
        data[6] = 0x50;
        data[7] = 0x50;

        sendFrame(0x0090, data, 8);

        std::cout << "   >>> 0x90流指令: Ang=" << angle_deg << "° Vel=" << max_vel_rps
                  << " Tqe=" << max_tqe_nm << " (原始值: [" << pos_int << "," << vel_int
                  << "," << tqe_int << "])" << std::endl;
    }

    void setAngle(double angle_deg, double max_vel_rps = 2.0, double max_tqe_nm = 3.0, int send_count = 5) {
        for (int i = 0; i < send_count; ++i) {
            send0x90Command(angle_deg, max_vel_rps, max_tqe_nm);
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
        std::cout << "   -> 目标角度: " << angle_deg << " 度" << std::endl;
    }

    void runInteractiveControl() {
        std::cout << "\n" << std::string(50, '=') << std::endl;
        std::cout << "🚀 0x90 流控制模式 (复刻 SDK)" << std::endl;
        std::cout << "输入角度 (如 90) 回车。" << std::endl;
        std::cout << "默认参数: 限速 2.0 r/s, 限矩 3.0 Nm" << std::endl;
        std::cout << "输入 q 退出" << std::endl;
        std::cout << std::string(50, '=') << std::endl;

        try {
            // 先使能电机
            enableMotor();

            while (g_running) {
                std::cout << "(Stream 0x90) > " << std::flush;

                // 检查输入流是否有效
                if (!std::cin.good()) {
                    break;
                }

                std::string input;
                std::getline(std::cin, input);

                if (input == "q" || input == "exit") {
                    break;
                }
                if (input.empty()) {
                    continue;
                }

                try {
                    double deg = std::stod(input);
                    setAngle(deg);
                } catch (...) {
                    std::cout << "输入错误" << std::endl;
                }
            }
        } catch (...) {
            std::cout << "\n中断" << std::endl;
        }
    }

    void runSineWave(double amplitude_deg, double frequency_hz, double duration_sec) {
        std::cout << "\n" << std::string(50, '=') << std::endl;
        std::cout << "🌊 正弦波角度控制" << std::endl;
        std::cout << "幅值: " << amplitude_deg << "°, 频率: " << frequency_hz
                  << " Hz, 时长: " << duration_sec << "s" << std::endl;
        std::cout << std::string(50, '=') << std::endl;

        try {
            enableMotor();

            auto start_time = std::chrono::steady_clock::now();
            while (g_running && std::chrono::duration_cast<std::chrono::seconds>(
                                    std::chrono::steady_clock::now() - start_time).count() < duration_sec) {

                auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
                                 std::chrono::steady_clock::now() - start_time).count() / 1000.0;
                double target_deg = amplitude_deg * std::sin(2.0 * M_PI * frequency_hz * elapsed);

                setAngle(target_deg);

                std::cout << "\r目标: " << std::fixed << std::setprecision(1) << target_deg << "°" << std::flush;
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
            }

        } catch (...) {
            std::cout << "\n中断" << std::endl;
        }
    }

    void runStepControl(const std::vector<double>& angles, double step_duration_sec) {
        std::cout << "\n" << std::string(50, '=') << std::endl;
        std::cout << "📈 阶梯角度控制" << std::endl;
        std::cout << "角度序列: ";
        for (size_t i = 0; i < angles.size(); ++i) {
            std::cout << angles[i];
            if (i < angles.size() - 1) std::cout << ", ";
        }
        std::cout << "°" << std::endl;
        std::cout << "每步时长: " << step_duration_sec << "s" << std::endl;
        std::cout << std::string(50, '=') << std::endl;

        try {
            enableMotor();

            for (size_t i = 0; i < angles.size() && g_running; ++i) {
                std::cout << "\n--- 步骤 " << (i + 1) << "/" << angles.size() << ": "
                          << angles[i] << "° ---" << std::endl;

                setAngle(angles[i]);

                auto step_start = std::chrono::steady_clock::now();
                while (g_running && std::chrono::duration_cast<std::chrono::seconds>(
                                        std::chrono::steady_clock::now() - step_start).count() < step_duration_sec) {

                    auto remaining = step_duration_sec -
                                   std::chrono::duration_cast<std::chrono::seconds>(
                                       std::chrono::steady_clock::now() - step_start).count();
                    std::cout << "\r剩余时间: " << std::fixed << std::setprecision(1)
                              << remaining << "s" << std::flush;
                    std::this_thread::sleep_for(std::chrono::milliseconds(100));
                }
            }

        } catch (...) {
            std::cout << "\n中断" << std::endl;
        }
    }

    void testPositions(const std::vector<double>& positions) {
        std::cout << "\n" << std::string(50, '=') << std::endl;
        std::cout << "🧪 多位置测试" << std::endl;
        std::cout << "测试位置: ";
        for (size_t i = 0; i < positions.size(); ++i) {
            std::cout << positions[i];
            if (i < positions.size() - 1) std::cout << ", ";
        }
        std::cout << "°" << std::endl;
        std::cout << std::string(50, '=') << std::endl;

        try {
            enableMotor();

            for (size_t i = 0; i < positions.size() && g_running; ++i) {
                std::cout << "\n--- 测试位置 " << (i + 1) << "/" << positions.size()
                          << ": " << positions[i] << "° ---" << std::endl;

                setAngle(positions[i]);

                std::cout << "等待2秒稳定..." << std::flush;
                std::this_thread::sleep_for(std::chrono::seconds(2));
            }

        } catch (...) {
            std::cout << "\n中断" << std::endl;
        }
    }
};

void showHelp() {
    std::cout << "LivelyBot 角度流控制" << std::endl;
    std::cout << "用法:" << std::endl;
    std::cout << "  ./angle_stream_control [motor_id] [mode] [options]" << std::endl;
    std::cout << "参数:" << std::endl;
    std::cout << "  motor_id  电机ID (默认: 1)" << std::endl;
    std::cout << "  mode      控制模式 (interactive, sine, step, test)" << std::endl;
    std::cout << "模式选项:" << std::endl;
    std::cout << "  sine:     --amplitude <度数> --frequency <Hz> --duration <秒>" << std::endl;
    std::cout << "  step:     --angles <角度列表,逗号分隔> --step-time <秒>" << std::endl;
    std::cout << "  test:     --positions <位置列表,逗号分隔>" << std::endl;
    std::cout << "示例:" << std::endl;
    std::cout << "  ./angle_stream_control 1 interactive" << std::endl;
    std::cout << "  ./angle_stream_control 1 sine --amplitude 90 --frequency 0.2 --duration 10" << std::endl;
    std::cout << "  ./angle_stream_control 1 step --angles \"0,45,90,45,0\" --step-time 3" << std::endl;
}

std::vector<double> parseDoubleList(const std::string& str) {
    std::vector<double> result;
    std::stringstream ss(str);
    std::string token;
    while (std::getline(ss, token, ',')) {
        result.push_back(std::stod(token));
    }
    return result;
}

int main(int argc, char* argv[]) {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    std::cout << "LivelyBot High Torque Angle Control" << std::endl;
    std::cout << "===================================" << std::endl;

    if (argc > 1 && (std::string(argv[1]) == "--help" || std::string(argv[1]) == "-h")) {
        showHelp();
        return 0;
    }

    int motor_id = 1;
    std::string mode = "interactive";
    double amplitude = 90.0;
    double frequency = 0.2;
    double duration = 10.0;
    std::vector<double> angles = {0.0, 45.0, 90.0, 45.0, 0.0};
    double step_time = 3.0;
    std::vector<double> positions = {0.0, 30.0, 60.0, 90.0, 60.0, 30.0, 0.0};

    // 解析命令行参数
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--motor-id" && i + 1 < argc) {
            motor_id = std::atoi(argv[++i]);
        } else if (arg == "--amplitude" && i + 1 < argc) {
            amplitude = std::stod(argv[++i]);
        } else if (arg == "--frequency" && i + 1 < argc) {
            frequency = std::stod(argv[++i]);
        } else if (arg == "--duration" && i + 1 < argc) {
            duration = std::stod(argv[++i]);
        } else if (arg == "--angles" && i + 1 < argc) {
            angles = parseDoubleList(argv[++i]);
        } else if (arg == "--step-time" && i + 1 < argc) {
            step_time = std::stod(argv[++i]);
        } else if (arg == "--positions" && i + 1 < argc) {
            positions = parseDoubleList(argv[++i]);
        } else if (arg.substr(0, 2) != "--") {
            if (i == 1) motor_id = std::atoi(arg.c_str());
            else if (i == 2) mode = arg;
        }
    }

    SimpleAngleController controller("can0", 1000000, motor_id);

    if (!controller.initialize()) {
        std::cerr << "控制器初始化失败" << std::endl;
        return 1;
    }

    try {
        if (mode == "interactive") {
            controller.runInteractiveControl();
        } else if (mode == "sine") {
            controller.runSineWave(amplitude, frequency, duration);
        } else if (mode == "step") {
            controller.runStepControl(angles, step_time);
        } else if (mode == "test") {
            controller.testPositions(positions);
        } else {
            std::cerr << "未知模式: " << mode << std::endl;
            showHelp();
            return 1;
        }
    } catch (const std::exception& e) {
        std::cout << "\n❌ 控制过程中出错: " << e.what() << std::endl;
        return 1;
    }

    std::cout << "\n程序结束" << std::endl;
    return 0;
}